Implementation of registered artists for the plotting framework (#874)
compas-dev/compas: docs/tutorial/plotters.rst

**************
Plotters
**************

> The COMPAS plotters (:mod:`compas_plotters`) provide an easy-to-use inteface for basic 2D visualisation

Reviewer: ```suggestion
The COMPAS plotters (:mod:`compas_plotters`) provide an easy-to-use interface for basic 2D visualization
```

of COMPAS objects based on matplotlib.

The package contains four types of plotters: :class:`compas_plotters.GeometryPlotter`, :class:`compas_plotters.NetworkPlotter`, :class:`compas_plotters.MeshPlotter`, and ... :class:`compas_plotters.Plotter`.

> The first three are deprecated in favour of :class:`compas_plotters.Plotter`, which is therefore the only one that will be described in this tutorial.

Reviewer: ```suggestion
The first three are deprecated in favor of :class:`compas_plotters.Plotter`, which is therefore the only one that will be described in this tutorial.
```

    plotter.add_from_list(cloud.points, size=1, facecolor=(1.0, 0.7, 0.7), edgecolor=(1.0, 0, 0))

Geometry Objects
================

Most of the geometry primitives are supported
and can be added to a plotter instance as described above:

* :class:`compas.geometry.Point`
* :class:`compas.geometry.Vector`
* :class:`compas.geometry.Line`
* :class:`compas.geometry.Circle`
* :class:`compas.geometry.Ellipse`
* :class:`compas.geometry.Polyline`
* :class:`compas.geometry.Polygon`

Bezier curves and pointclouds are currently not available yet, but will be added as well.
Note that in all cases, the ``z`` coordinates of the objects are simply ignored, and only a 2D representation is depicted.

.. code-block:: python

    plotter.add(point)
    plotter.add(vector)
    plotter.add(line)
    plotter.add(circle)
    plotter.add(ellipse)
    plotter.add(polyline)
    plotter.add(polygon)

Data Structures
===============

Of the three types of data structures, only network and mesh are supported.
Also in this case, the ``z`` coordinates of the geometry is ignored, and only a 2D representation is depicted.

.. code-block:: python

    plotter.add(point)
    plotter.add(vector)

> Visualisation Options

Reviewer: ```suggestion
Visualization Options
```

=====================

Line and Polyline
-----------------

.. rst-class:: table table-bordered

.. list-table::
    :widths: auto
    :header-rows: 1

    * - Name
      - Value
      - Default
    * - ``linewidth``
      - :obj:`float`
      - ``1.0``
    * - ``linestyle``
      - ``{'solid', 'dotted', 'dashed', 'dashdot'}``
      - ``'solid'``
    * - ``color``
      - :obj:`tuple`
      - ``(0.0, 0.0, 0.0)``
    * - ``draw_points``
      - :obj:`bool`
      - ``False``

.. code-block:: python

    pointcloud = Pointcloud.from_bounds(8, 5, 0, 10)

>     for a, b in grouper(pointcloud, 2):

Reviewer: Is this `compas.utilities.grouper`?

    pointcloud = Pointcloud.from_bounds(8, 5, 0, 10)

    for point in pointcloud:
        plotter.add(point, size=random.randint(1, 10), edgecolor=i_to_rgb(random.random(), normalize=True))

.. figure:: /_images/tutorial/plotters_point-options.png
    :figclass: figure
    :class: figure-img img-fluid

Vectors
-------

.. rst-class:: table table-bordered

.. list-table::
    :widths: auto
    :header-rows: 1

    * - Name
      - Value
      - Default
    * - ``point``
      - :class:`compas.geometry.Point`
      - ``None``
    * - ``draw_point``
      - :obj:`bool`
      - ``False``
    * - ``color``
      - :obj:`tuple`
      - ``(0.0, 0.0, 0.0)``

.. code-block:: python

    pointcloud = Pointcloud.from_bounds(8, 5, 0, 10)

    for index, (a, b) in enumerate(pairwise(pointcloud)):
        vector = b - a
        vector.unitize()
        plotter.add(vector, point=a, draw_point=True, color=i_to_red(max(index / 10, 0.1), normalize=True))

.. figure:: /_images/tutorial/plotters_vector-options.png
    :figclass: figure
    :class: figure-img img-fluid

Dynamic Plots
=============

Dynamic plots, or animations, can be made with the "on" decorator :meth:`compas_plotters.Plotter.on`.

> Simply add the decorator to a callback functions that updates the geometry in the plot at a specified interval.

Reviewer: ```suggestion
Simply add the decorator to a callback function that updates the geometry in the plot at a specified interval.
```
